crates/http/src/server/mod.rs

```diff
@@ -0,0 +1,165 @@
+///
+/// Create a new server using the [`ServerBuilder`](builder::ServerBuilder) struct.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct HttpServer<S> {
+    ctx: scuffle_context::Context,
+    service_factory: S,
+    bind: SocketAddr,
+    #[cfg(feature = "http1")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "http1")))]
+    enable_http1: bool,
+    #[cfg(feature = "http2")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "http2")))]
+    enable_http2: bool,
+    #[cfg(feature = "http3")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "http3")))]
+    enable_http3: bool,
+    #[cfg(feature = "tls-rustls")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "tls-rustls")))]
+    rustls_config: Option<rustls::ServerConfig>,
+}
+
+impl<F> HttpServer<F>
+where
+    F: HttpServiceFactory,
+{
+    /// Entry point for creating a new HTTP server.
+    pub fn builder() -> builder::ServerBuilder<(), F> {
+        builder::ServerBuilder::default()
+    }
+}
+
+impl<F> HttpServer<F>
+where
+    F: HttpServiceFactory + Clone + Send + 'static,
+    F::Error: Debug + Display,
+    F::Service: Clone + Send + 'static,
+    <F::Service as HttpService>::Error: std::error::Error + Debug + Display + Send + Sync,
+    <F::Service as HttpService>::ResBody: Send,
+    <<F::Service as HttpService>::ResBody as http_body::Body>::Data: Send,
+    <<F::Service as HttpService>::ResBody as http_body::Body>::Error: std::error::Error + Send + Sync,
+{
+    /// Run the server.
+    ///
+    /// This will:
+    ///
+    /// - Start listening on all configured interfaces for incoming connections.
+    /// - Accept all incoming connections.
+    /// - Handle incoming requests by passing them to the configured service factory.
+    pub async fn run(self) -> Result<(), Error<F>> {
+        #[allow(unused_variables)]
+        #[cfg(not(any(feature = "http1", feature = "http2")))]
+        let start_tcp_backend = false;
+        #[allow(unused_variables)]
+        #[cfg(feature = "http1")]
+        let start_tcp_backend = self.enable_http1;
+        #[allow(unused_variables)]
+        #[cfg(feature = "http2")]
+        let start_tcp_backend = self.enable_http2;
+        #[cfg(all(feature = "http1", feature = "http2"))]
+        let start_tcp_backend = self.enable_http1 || self.enable_http2;
+
+        #[cfg(feature = "tls-rustls")]
+        if let Some(_rustls_config) = self.rustls_config {
+            #[allow(unused_variables)]
+            let enable_http3 = false;
+            #[cfg(feature = "http3")]
+            let enable_http3 = self.enable_http3;
+
+            match (start_tcp_backend, enable_http3) {
+                #[cfg(feature = "http3")]
+                (false, true) => {
+                    use scuffle_context::ContextFutExt;
+
+                    let backend = crate::backend::h3::Http3Backend { bind: self.bind };
+
+                    match backend.run(self.service_factory, _rustls_config).with_context(self.ctx).await {
```

Comment:
Here we add with_context but we never send the context into the future and therefore any Tokio spawns will not be bound by the context.